third_party/move/move-compiler-v2/src/cyclic_instantiation_checker.rs

```diff
@@ -0,0 +1,242 @@
+            true
+        }
+    }
+
+    /// Visits a call expression and checks for cyclic type instantiations.
+    /// Other parameters are the same as in `visit`.
+    /// Precondition: `caller` defined in `self.mod_env`
+    fn visit_callees(
+        &self,
+        caller_node: NodeId,
+        caller: QualifiedInstId<FunId>,
+        insts: &[Type],
+        callers_chain: &mut Vec<(Loc, QualifiedInstId<FunId>)>,
+    ) -> bool {
+        let fun_env = self.mod_env.get_function(caller.id);
+        if let Some(caller_body) = fun_env.get_def() {
+            let caller_loc = self.mod_env.env.get_node_loc(caller_node);
+            callers_chain.push((caller_loc, caller));
+            let insts = Type::instantiate_vec(self.get_inst(caller_node), insts);
+            let res = caller_body.visit_positions_with_return_val(&mut |pos, exp| {
+                self.visit(pos, exp, &insts, callers_chain)
+            });
+            callers_chain.pop();
+            res
+        } else {
+            true
+        }
+    }
+
+    /// Shortcut for getting the node instantiation
+    fn get_inst(&self, nid: NodeId) -> Vec<Type> {
+        self.mod_env
+            .env
+            .get_node_instantiation_opt(nid)
+            .expect("instantiation")
+    }
+
+    /// Returns true if we are sure the function with given id is not recursive
+    fn def_not_recursive(&self, id: QualifiedId<FunId>) -> bool {
+        if let Some(descendants) = self
+            .mod_env
+            .env
+            .get_function(id)
+            .get_transitive_closure_of_called_functions()
+        {
+            !descendants.contains(&id)
+        } else {
+            false
+        }
+    }
+
+    /// Reports a cyclic type instantiation error, in which the root caller eventually calls `callee`
+    /// with a cyclic type instantiation. `callee` is the callee of the last caller in `callers_chain`.
+    fn report_error(
+        &self,
+        _nid: NodeId,
+        callee: QualifiedInstId<FunId>,
+        callers_chain: &mut Vec<(Loc, QualifiedInstId<FunId>)>,
+    ) {
+        let labels = (0..callers_chain.len())
+            .map(|i| {
+                let (caller_loc, caller) = &callers_chain[i];
+                // callee of `caller`
+                let callee = if i != callers_chain.len() - 1 {
+                    &callers_chain[i + 1].1
+                } else {
+                    // callee of the last caller in the chain is the parameter `callee`
+                    &callee
+                };
+                format!(
+                    "{} calls {} {}",
+                    self.display_call(caller),
+                    self.display_call(callee),
+                    caller_loc.display_line_only(self.mod_env.env)
+                )
+            })
+            .collect_vec();
+        let root_loc = &callers_chain[0].0;
+        self.mod_env
+            .env
+            .error_with_notes(root_loc, "cyclic type instantiation", labels)
```

Comment:
This error needs more explanation.  V1 says "cyclic type instantiation" but adds a note: "The type parameter 'c::T1' was instantiated with the type '0x8675309::M::S<T2>', which contains the type parameter 'c::T2'. A cycle of recursive calls causes the instantiation to recurse infinitely".  The V1 explanation makes me feel like I understand the problem.

Reply:
Added "A cycle of recursive calls causes the instantiation to recurse infinitely". But I didn't until the first part of that error message till now; it's rather confusing.